src/Microsoft.ApiDesignGuidelines.Analyzers/Core/EnumsShouldHavePluralNames.cs

```diff
@@ -0,0 +1,132 @@
+                MicrosoftApiDesignGuidelinesAnalyzersResources.ResourceManager,
+                typeof(MicrosoftApiDesignGuidelinesAnalyzersResources));
+
+        private static readonly LocalizableString s_localizableMessage_CA1717 =
+            new LocalizableResourceString(
+                nameof(MicrosoftApiDesignGuidelinesAnalyzersResources.OnlyFlagsEnumsShouldHavePluralNamesMessage),
+                MicrosoftApiDesignGuidelinesAnalyzersResources.ResourceManager,
+                typeof(MicrosoftApiDesignGuidelinesAnalyzersResources));
+
+        private static readonly LocalizableString s_localizableDescription_CA1717 =
+            new LocalizableResourceString(
+                nameof(MicrosoftApiDesignGuidelinesAnalyzersResources.OnlyFlagsEnumsShouldHavePluralNamesDescription),
+                MicrosoftApiDesignGuidelinesAnalyzersResources.ResourceManager,
+                typeof(MicrosoftApiDesignGuidelinesAnalyzersResources));
+
+        internal static DiagnosticDescriptor Rule_CA1717 =
+            new DiagnosticDescriptor(
+                RuleId_NoPlural,
+                s_localizableTitle_CA1717,
+                s_localizableMessage_CA1717,
+                DiagnosticCategory.Naming,
+                DiagnosticSeverity.Warning,
+                isEnabledByDefault: true,
+                description: s_localizableDescription_CA1717,
+                helpLinkUri: "https://msdn.microsoft.com/en-us/library/bb264487.aspx",
+                customTags: WellKnownDiagnosticTags.Telemetry);
+
+        #endregion
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule_CA1714, Rule_CA1717);
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.RegisterCompilationStartAction(compilationContext =>
+            {
+                var flagsAttribute = WellKnownTypes.FlagsAttribute(compilationContext.Compilation);
+                if (flagsAttribute == null)
+                {
+                    return;
+                }
+
+                compilationContext.RegisterSymbolAction(symbolContext => AnalyzeSymbol(symbolContext, flagsAttribute), SymbolKind.NamedType);
+            });
+
+        }
+
+        private static void AnalyzeSymbol(SymbolAnalysisContext context, INamedTypeSymbol flagsAttribute)
+        {
+            var symbol = (INamedTypeSymbol)context.Symbol;
+            if (symbol.TypeKind != TypeKind.Enum)
+            {
+                return;
+            }
+
+            bool hasFlagsAttribute = symbol.GetAttributes().Any(a => a.AttributeClass.Equals(flagsAttribute));
+            if (hasFlagsAttribute)
+            {
+                if (!symbol.Name.IsPlural()) // Checking Rule CA1714
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(Rule_CA1714, symbol.Locations.First(), symbol.Name));
```

Comment:
You can write symbol.CreateDiagnostic - that creates the diagnostic in the first non-metadata location.